packages/integration-sdk-http-client/src/index.ts

```diff
@@ -19,77 +14,86 @@ export type APIResourceIterationCallback<T> = (
+    return true;
+  } else {
+    return false;
+  }
+};
+
+export type APIErrorHandlerFunction = (
+  request: APIRequest,
+  response: Response,
+) => Promise<void> | void;
+
+const defaultErrorHandler: APIErrorHandlerFunction = async (
+  request,
+  response,
+) => {
+  if (response.status === 429) {
+    await handleRateLimitError(response.headers, DEFAULT_RATE_LIMIT_CONFIG);
+  } else if (
+    (response.status >= 400 && response.status !== 408) ||
+    response.status === 501
+  ) {
+    throw new APIError({
+      message: `${request.method} ${request.url}: ${response.status} ${response.statusText}`,
+      status: response.status,
+      statusText: response.statusText,
+      endpoint: request.url,
+    });
+  }
+};
+
 export class APIClient {
   // TODO (adam-in-ict) make rateLimitConfig configurable and only use the default values when none are provided
-  private rateLimitConfig: RateLimitConfig = DEFAULT_RATE_LIMIT_CONFIG;
-  private rateLimitState: RateLimitState;
 
-  async executeAPIRequest(request: APIRequest): Promise<APIResponse> {
+  async executeAPIRequest(
+    request: APIRequest,
+    errorHandler?: APIErrorHandlerFunction,
+    isRetryable?: isRetryableFunction,
+  ): Promise<APIResponse> {
     let attempts = 0;
 
-    do {
-      const response = await fetch(request.url, request);
+    errorHandler =
+      errorHandler === undefined ? defaultErrorHandler : errorHandler;
 
-      if (isNonRetryableError(response.status)) {
-        const err = new APIError({
-          message: `API request error for ${request.url}: ${response.status} ${response.statusText}`,
-          code: response.status,
-        });
-        throw err;
-      }
+    isRetryable = isRetryable === undefined ? defaultIsRetryable : isRetryable;
 
-      // TODO (adam-in-ict) make rate limit error configurable for instances where an API isn't using the default of 429
-      if (response.status === 429) {
-        await this.handleRateLimitError(response.headers);
-      } else {
-        // Successful response, so return
+    let response: Response;
+    do {
+      response = await fetch(request.url, request);
+
+      if (response.ok) {
         return {
+          // Will we always get json data?
           data: await response.json(),
           headers: response.headers,
           status: response.status,
           statusText: response.statusText,
         };
+      } else {
+        await errorHandler(request, response);
       }
-
       attempts += 1;
-    } while (attempts < this.rateLimitConfig.maxAttempts);
+    } while (isRetryable(attempts, request, response));
 
     throw new APIError({
       message: `Could not complete request within ${attempts} attempts!`,
-      code: 429,
+      status: response.status,
+      statusText: response.statusText,
+      endpoint: request.url,
     });
   }
-
-  //TODO (adam-in-ict) abstract out which Header values we're grabbing for rate limiting.
```

Comment:
Would we be able to leave this TODO in the code as well?  I'm not sure that the headers we're looking for will always be the case (I'm not actually even totally confident I selected the ones that are the most common).

Reply:
Yep. I should have brought it over. I'll add it back in!